File: android/app/src/main/java/quarky/com/br/mercuryjacket/ui/dialog/Alert.java
package quarky.com.br.mercuryjacket.ui.dialog;

import android.app.Activity;
import android.app.Dialog;
import android.content.Context;
import android.text.SpannableStringBuilder;
import android.view.View;
import android.view.Window;
import android.widget.TextView;

import quarky.com.br.mercuryjacket.R;

/**
 * Created by andreluisponce on 26/12/17.
 */

public class Alert {
    public static final int CONFIRM = 0;
    public static final int ASK = 1;
    public static final int WAIT = 2;

    public static final int CONFIRM_ACTION = 0;
    public static final int YES_ACTION = 1;
    public static final int NO_ACTION = 2;

    public String confirmMessage = "Continue";
    public String yesMessage = "Yes";
    public String noMessage = "No";

    private Context context;
    private ActionListener listener;
    private Dialog dialog;

    public static Alert INSTANCE;

    public Alert(Context context)
    {
        if(INSTANCE!=null) INSTANCE.dismiss();
        INSTANCE = this;
        this.listener = null;
        this.context = context;
    }

    public static Alert show(Context context, String message)
    {
        Alert alert = new Alert(context);
        alert.show(message,Alert.CONFIRM);
        return alert;
    }

    public Dialog show(String message, int ask) {
        SpannableStringBuilder newMessage = new SpannableStringBuilder();
        newMessage.append(message);
        return show(newMessage,ask);
    }

    public interface ActionListener {
        public void onActionClick(int action);
    }

    public void setActionListener(ActionListener listener) {
        this.listener = listener;
    }

    public Dialog show(final SpannableStringBuilder message, final int type)
    {
        Activity activity = (Activity) context;
        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                dialog = new Dialog(context, R.style.NewDialog);
                dialog.requestWindowFeature(Window.FEATURE_NO_TITLE);
                dialog.setCancelable(false);
                dialog.setContentView(R.layout.alert);

                View buttons_container = dialog.findViewById(R.id.buttons_container);
                View bt_ok             = dialog.findViewById(R.id.bt_ok);
                View bts_question      = dialog.findViewById(R.id.bts_question);
                View bt_yes            = dialog.findViewById(R.id.bt_yes);
                View bt_no             = dialog.findViewById(R.id.bt_no);

                switch(type)
                {
                    case CONFIRM:
                        buttons_container.setVisibility(View.VISIBLE);
                        bt_ok.setVisibility(View.VISIBLE);
                        bts_question.setVisibility(View.GONE);
                        break;
                    case ASK:
                        buttons_container.setVisibility(View.VISIBLE);
                        bt_ok.setVisibility(View.GONE);
                        bts_question.setVisibility(View.VISIBLE);
                        break;
                    case WAIT:
                        buttons_container.setVisibility(View.GONE);
                        break;
                }


                TextView txt_ok = bt_ok.findViewById(R.id.label_txt);
                TextView txt_yes = bt_yes.findViewById(R.id.label_txt);
                TextView txt_no = bt_no.findViewById(R.id.label_txt);
                TextView message_txt = (TextView) dialog.findViewById(R.id.message_txt);

                message_txt.setText(message);
                txt_ok.setText(confirmMessage);
                txt_yes.setText(yesMessage);
                txt_no.setText(noMessage);

                bt_ok.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View view) {
                        actionHandle(CONFIRM_ACTION);
                    }
                });

                bt_yes.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View view) {
                        actionHandle(YES_ACTION);
                    }
                });

                bt_no.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View view) {
                        actionHandle(NO_ACTION);
                    }
                });
            }
        });

        try {
            if(dialog!=null) {
                dialog.dismiss();
                dialog.show();
            }
        } catch (Exception e) {
            e.printStackTrace();
        }

        return dialog;
    }

    private void actionHandle(int action)
    {
        dismiss();
        try {
            listener.onActionClick(action);
            listener = null;
        }catch(NullPointerException e){

        }
    }

    public void dismiss()
    {
        dialog.dismiss();
        INSTANCE = null;
    }
}

File: ios-v2/Mercury Jacket/ui/dialog/Alert.swift
//
//  Alert.swift
//  Desafieme
//
//  Created by André Ponce on 16/10/2018.
//  Copyright © 2018 Quarky. All rights reserved.
//
import UIKit
import Foundation

class Alert
{
    private static var INSTANCE :Alert? = nil;
    
    public static let CONFIRM :Int = 0;
    public static let ASK     :Int = 1;
    public static let WAIT    :Int = 2;
    public static let LOADER  :Int = 3;
    
    public static let CONFIRM_ACTION :Int = 0;
    public static let YES_ACTION     :Int = 1;
    public static let NO_ACTION      :Int = 2;
    
    public var confirmMessage :String = "ok";
    public var yesMessage     :String = "Yes";
    public var noMessage      :String = "No";
    
    private var context :UIViewController;
    private var listener :ActionListener? = nil;
    private var alertController :UIAlertController? = nil;
    
    init(context :UIViewController){
        self.context = context;
        if(Alert.INSTANCE != nil) {
            Alert.INSTANCE?.dismiss();
        }
        Alert.INSTANCE = self;
    }
    
    class ActionListener{
        //public func onActionClick(action :Int) {}
        var onActionClick :((_ action :Int) -> Void);
        init(onActionClick :@escaping (_ action :Int) -> Void)
        {
            self.onActionClick = onActionClick;
        }
    }
    
    public func setActionListener(listener :ActionListener) {
        self.listener = listener;
    }
    
    public static func show(context :UIViewController, message :String) -> Alert
    {
        let alert :Alert = Alert(context: context);
        alert.create(message: message,type: Alert.CONFIRM);
        alert.show();
        return alert;
    }
    
    public func create(message :String, type :Int)
    {
        self.alertController = UIAlertController(title: nil, message: message, preferredStyle: UIAlertController.Style.alert);
        
        switch(type)
        {
        case Alert.CONFIRM:
            self.addAction(title: self.confirmMessage, code: Alert.CONFIRM_ACTION);
            break;
        case Alert.ASK:
            self.addAction(title: self.yesMessage, code: Alert.YES_ACTION);
            self.addAction(title: self.noMessage, code: Alert.NO_ACTION);
            break;
        default: break
        }
    }
    
    private func addAction(title :String, code :Int)
    {
        self.alertController?.addAction(UIAlertAction(title: title, style: code == Alert.NO_ACTION ? UIAlertAction.Style.cancel : UIAlertAction.Style.default, handler: { (action) in
            self.listener?.onActionClick(code)
        }))
    }
    
    public func show()
    {
        if(self.alertController != nil) {
            print("SHOW ALERT");
            self.context.present(self.alertController!, animated: true, completion: nil)
        }
    }
    
    public func dismiss()
    {
        if(self.alertController != nil)
        {
            print("DISMISS ALERT");
            self.alertController?.removeFromParent()
            self.alertController?.dismiss(animated: false, completion: nil)
            self.alertController = nil
        }
    }
}
